sol-did/programs/sol-did/src/lib.rs

```diff
@@ -93,3 +96,317 @@ pub mod sol_did {
         instructions::migrate(ctx)
     }
 }
+
+// /// Given a DidAccount, and a list of controlling did accountInfos,
+// /// Parse the controlling did accountInfos into DidAccounts,
+// /// and validate that the controller chain is correct
+// fn valid_controller_chain<'a, 'b: 'a>(
+//     did_data: &'a Account<DidAccount>,
+//     did_accounts: &'b [AccountInfo<'b>],
+// ) -> Result<Vec<Account<'b, DidAccount>>> {
+//     let controller_chain: Vec<Account<DidAccount>> = did_accounts
+//         .iter()
+//         .map(Account::try_from)
+//         .collect::<Result<Vec<Account<DidAccount>>>>()?;
+//
+//     if !did_data.is_controlled_by(controller_chain.as_slice()) {
+//         return Err(error!(DidSolError::InvalidControllerChain));
+//     }
+//
+//     Ok(controller_chain)
+// }
+//
+// fn last_in_valid_controller_chain<'a, 'b: 'a>(
+//     did_data: &'a Account<DidAccount>,
+//     did_accounts: &'b [AccountInfo<'b>],
+// ) -> Result<Option<&'b Account<'b, DidAccount>>> {
+//     valid_controller_chain(did_data, did_accounts).map(|chain| chain.last())
+// }
+
+/// Will return if given solana public key, or ethereum address (as derived from an ethereum signature)
+/// is a valid authority (CAPABILITY_INVOCATION) on the given did_account.
+/// Authorities may be direct, or via a chain of controlling did accounts.
+/// In the latter case, the chain must be provided in the following order:
+/// did_account -> controlling_did_accounts[0] -> ... -> controlling_did_accounts[n] -> authority
+/// where '->' represents the relationship "is controlled by".
+pub fn is_authority(
+    did_account: &AccountInfo,
+    controlling_did_accounts: &[AccountInfo],
+    sol_authority: &Pubkey,
+    eth_message: &[u8],
+    eth_raw_signature: Option<&Secp256k1RawSignature>,
+    filter_fragment: Option<&String>,
+) -> Result<bool> {
+    if did_account.owner == &System::id() && did_account.lamports() == 0 {
+        msg!("Validating generative DID");
+        // the DID is a generative DID - the only authority is the key itself
+        // verify that the authority key derives the correct did account
+        let address = derive_did_account(sol_authority).0;
```

Comment:
`Pubkey::find_program_address` is expensive. I think there is a cheaper way to verify derivation if all inputs are known. I'll check the anchor code for that.

Reply:
```
As all account addresses accessed by an on-chain Solana program must be explicitly passed to the program, it is typical for the PDAs to be derived in off-chain client programs, avoiding the compute cost of generating the address on-chain. The address may or may not then be verified by re-deriving it on-chain, depending on the requirements of the program. This verification may be performed without the overhead of re-searching for the bump key by using the create_program_address function.
```

Reply:
Hm the problem is that for create_program_address I would need to pass in the bump to verify the generative case, which makes the interface less clean.

Reply:
@dankelleher I've changed the function to use `create_program_address` but that also requires a bump now.